cranelift/codegen/src/machinst/abi.rs

```diff
@@ -124,6 +124,17 @@ use std::mem;
 /// a small fixed sequence implementing one operation.
 pub type SmallInstVec<I> = SmallVec<[I; 4]>;
 
+/// A type used by backends to track argument-binding info in the
+/// "args" pseudoinst.
+#[derive(Clone, Debug)]
+pub struct ArgInfo {
+    /// vregs that are defined at the top of the function with
+    /// register-argument values.
+    pub defs: Vec<Writable<Reg>>,
+    /// pregs that constrain these vregs to ABI-defined locations.
+    pub pregs: Vec<Reg>,
+}
```

Comment:
Every user of `ArgInfo` currently has to zip together `defs` and `pregs`. I try to avoid that where possible because it means there's an invariant that the two vectors have the same length, which has to be maintained across distant parts of the codebase. I'd suggest instead either using a vector of pairs, or if you want to name the elements of the pair to make it clear which is which, then have this struct represent only one constraint and keep a `Vec<ArgInfo>` around instead.

```suggestion
pub struct ArgInfo {
    /// vreg defined at the top of the function with
    /// register-argument values.
    pub def: Writable<Reg>,
    /// preg that constrains that vreg to ABI-defined locations.
    pub preg: Reg,
}
```

Reply:
I really like the suggestion to have the args be a vector of pairs 👍

Reply:
Done; this is much cleaner now, thanks!